### ISSUE TITLE
Improve the typing of `foreach`

### ISSUE THREAD

**Issue by MatthewDaggitt** [vehicle-lang/vehicle#114]
Follow up from #83. At the moment the type of `foreach ... in...` is:
```haskell
  forall tNat $ \d ->
    forall type0 $ \tElem ->
      let ds = lseq tNat (tList tNat) [d] in
      (tElem ~> tBool) ~> tTensor tElem ds ~> tTensor tBool ds
```
which is a hack. Really we should be completely polymorphic in the type of container we iterate over. Unfortunately we can't do this properly until we have a proper, `IsContainer f` type class where `f` is a type-level function, mapping elements to containers of elements.

**Comment by MatthewDaggitt** [comment]
@bobatkey has pointed out that we should make `foreach` be a general constructor for tensors, in a similar way to the `for` constructor in [Dex](https://openreview.net/pdf?id=rJxd7vsWPS). Makes a lot of sense.

**Comment by MatthewDaggitt** [comment]
Fixed in [c7dada8](https://github.com/vehicle-lang/vehicle/commit/c7dada8853bde6de5dc85b7f817d751aad291eb8)
